[RISCV][TTI] Implement getAddressComputationCost() in RISCV TTI.

ElvisWang123 opened the pull request:
This patch implements the `getAddressComputationCost()` in RISCV TTI which
make the gather/scatter with address calculation more expansive that
stride cost.

Note that the only user of `getAddressComputationCost()` with vector
type is in `VPWidenMemoryRecipe::computeCost()`. So this patch make some
LV tests changes.

I've checked the tests changes in LV and seems those changes can be
divided into two groups.
 * gather/scatter with uniform vector ptr, seems can be optimized to
 masked.load.
 * can optimize to stride load/store.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1085,6 +1085,12 @@ InstructionCost RISCVTTIImpl::getGatherScatterOpCost(
     return BaseT::getGatherScatterOpCost(Opcode, DataTy, Ptr, VariableMask,
                                          Alignment, CostKind, I);
 
+  // Gather/Scatter instruction will need to calculate the address of each
```
I don't know if I understand this right. Where do we need an address calculation?

ElvisWang123 replied:
Oh, yes. This should be implemented in `getAddressComputationCost()`.

ElvisWang123 committed:
[RISCV][TTI] Implement `getAddressComputationCost()` in RISCV TTI.

This patch impelemt the `getAddressComputationCost()` in RISCV TTI which
make the gather/scatter with address calculation more expansive that
stride cost.

Note that the only user of `getAddressComputationCost()` with vector
type is in `VPWidenMemoryRecipe::computeCost()`. So this patch make some
LV tests changes.

I've checked the tests changes in LV and seems those changes can be
divided into two groups.
 * gather/scatter with uniform vector ptr, seems can be optimized to
 masked.load.
 * can optimize to stirde load/store.

ElvisWang123 commented on llvm/test/Transforms/SLPVectorizer/RISCV/remarks-insert-into-small-vector.ll:
```diff
@@ -8,7 +8,7 @@
 ; YAML-NEXT:  Function:        test
 ; YAML-NEXT:  Args:
 ; YAML-NEXT:  - String:          'Stores SLP vectorized with cost '
-; YAML-NEXT:  - Cost:            '-2'
```
Move the address computation cost to `getAddressComputationCost()`.

ElvisWang123 reviewed (commented):
Move  the address computation cost to `getAddressComputationCost`. 

Some of the LV tests changes seems can prevent. 
I will try to optimize the gather/scatter with uniform address if possible.

lukel97 commented on llvm/test/Transforms/LoopVectorize/RISCV/blend-any-of-reduction-cost.ll:
```diff
@@ -62,50 +62,10 @@ define i32 @any_of_reduction_used_in_blend_with_multiple_phis(ptr %src, i64 %N,
-; CHECK:       [[VECTOR_BODY]]:
-; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, %[[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], %[[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[VEC_PHI:%.*]] = phi <vscale x 2 x i1> [ zeroinitializer, %[[VECTOR_PH]] ], [ [[PREDPHI:%.*]], %[[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[WIDE_MASKED_GATHER:%.*]] = call <vscale x 2 x ptr> @llvm.masked.gather.nxv2p0.nxv2p0(<vscale x 2 x ptr> [[BROADCAST_SPLAT4]], i32 8, <vscale x 2 x i1> [[TMP8]], <vscale x 2 x ptr> poison)
```
As you say, I think this should only be adding the scalar cost here because it's uniform. Is checking that the SCEV isn't an addrec good enough? I think we can use `!BaseT::isStridedAccess(Ptr)` for that?

ElvisWang123 replied:
Good point!
I think using  `!BaseT::isStridedAccess(Ptr)` need to change the `VPWIdenMemory::computeCost()` and rewrite some functions in vplan. 
 Currently `VPWidenMemoryRecipe` doesn't contain SCEV `ptr` and it hard to infer by the `ptr` in the underlying  instruction (`getAddressAccessSCEV()` needs `LoopVectorizationLegality`, `PredicatedScalarEvolution` and `Loop` which is not in the VPlan). 

I think check if the addr is uniform during `VPWidenMemory::computeCost` is better. If the addr is uniform, query  `getAddressComputationCost()` with scalar type.

lukel97 replied:
Should we do that first before landing this PR? Otherwise I think this test diff is a regression, it should be profitable to vectorize this loop but we're bailing now because the address computation cost is too high

ElvisWang123 replied:
Yes, I will fix `VPWIdenMemoryRecipe::computeCost()` first to prevent the regression in this PR.

lukel97 reviewed (commented):
I can't believe we haven't already implemented this, nice find.

wangpc-pp reviewed (approved):
Nice! LGTM!

topperc commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1503,6 +1503,18 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+InstructionCost RISCVTTIImpl::getAddressComputationCost(Type *Ty,
+                                                        ScalarEvolution *SE,
+                                                        const SCEV *Ptr) const {
+  // Address computations with vector type is for usually for index load/store
```
```suggestion
  // Address computations with vector type are usually for indexed load/store
```

ElvisWang123 replied:
Fixed, thanks!

ElvisWang123 reviewed (commented):
Base on updating `VPWidenMemoryRecipe::computeCost()` in LV #150371.

`vectorie-force-tail-with-evl-gather-scatter` looks like won't be vectorized since the cost of gather/scatter increase.

topperc commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3083,9 +3083,18 @@ InstructionCost VPWidenMemoryRecipe::computeCost(ElementCount VF,
+    InstructionCost Cost = 0;
+
+    // If the address value is uniform across all lane, then the address can be
+    // calculated with scalar type and broacast.
```
```suggestion
    // calculated with scalar type and broadcast.
```

ElvisWang123 commented:
Rebase on #150371 to show how test changes.

alexey-bataev commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1566,6 +1566,19 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+  // which is likely more expensive.
+  if (ST->hasVInstructions() && PtrTy->isVectorTy())
+    return getArithmeticInstrCost(
+        Instruction::Add, PtrTy, TTI::TCK_RecipThroughput,
```
Better to add a cost kind parameter to the function rather than hardcode it here.

ElvisWang123 replied:
Added in #153342. Thanks!

ElvisWang123 committed:
Update tests after ed52bdd

ElvisWang123 commented:
After #150371 landed, the potential regression is gone ([5db0a39](https://github.com/llvm/llvm-project/pull/149955/commits/5db0a39d39933bb46b225ce0eef5e03ebd7a4eef))

lukel97 reviewed (approved):
LGTM. I looked at the two loops in the diff and it does look like they were probably unprofitably vectorized. gather_scatter needs a vsll.vi to scale the offsets, and store_factor_4_reverse needs a vwmulsu.vx.

ElvisWang123 committed:
Address comments.

ElvisWang123 commented:
Reverted, will fix the assertion (cost-model disagree) in llvm-test-suite and try to landed again.

fhahn reviewed (commented):
Looks like there's another crash: https://llvm.godbolt.org/z/7dzxxzYcK

ElvisWang123 commented:
Thanks for reporting the issue. Will fix this ASAP
